### PR TITLE
K8SPSMDB-419: add concurrent jobs check using PBM locks

### PR DESCRIPTION
**Pull request by ollevche**
[K8SPSMDB-419](https://jira.percona.com/browse/K8SPSMDB-419) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPSMDB-419

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| self-healing | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | failed |
| version-service | passed |
| upgrade-sharded | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/609/commits/555cece9a08e5b637d2a7ebbccb0ddde116e3081
image: `perconalab/percona-server-mongodb-operator:PR-609-555cece9`

File: pkg/controller/perconaservermongodb/smart.go [hunk 1]
```diff
@@ -58,12 +59,12 @@ func (r *ReconcilePerconaServerMongoDB) smartUpdate(cr *api.PerconaServerMongoDB
 		return nil
 	}
 
-	ok, err := r.isBackupRunning(cr)
+	ok, err := backup.HasActiveJobs(r.client, cr, backup.Job{})
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/smart.go:62]: can we please don't completely remove the current `isBackupRunning` but in addition to `isRestoreRunning` and to `isBackupRunning` also run `HasActiveJobs` check?

File: pkg/psmdb/backup/backup.go [hunk 2]
```diff
@@ -60,13 +74,29 @@ func HasActiveJobs(cl client.Client, cluster, namespace string, current Job) (bo
...
+	hasLocks, err := pbm.HasLocks(allowLock...)
+	if err != nil {
+		return false, errors.Wrap(err, "getting pbm locks")
+	}
+
+	if hasLocks {
+		return true, nil
+	}
+
 	return false, nil
```
Comment (delgod) [inline pkg/psmdb/backup/backup.go:101]:
```suggestion
	return pbm.HasLocks(allowLock...)
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/psmdb/backup/backup.go (hunk 2 above)
Comment (ollevche) [reply pkg/psmdb/backup/backup.go:101]: Are you sure? It's shorter, but it's also harder to add something to the end of the function, isn't it?

File: pkg/controller/perconaservermongodb/smart.go (hunk 1 above)
Comment (ollevche) [reply pkg/controller/perconaservermongodb/smart.go]: added it back in dbb7b7b3177ca3df49dd8fbf78553d91eacdde74

File: pkg/controller/perconaservermongodb/smart.go [hunk 3]
```diff
@@ -3,6 +3,7 @@ package perconaservermongodb
 import (
 	"context"
 	"fmt"
+	"github.com/percona/percona-server-mongodb-operator/pkg/psmdb/backup"
```
Comment (qJkee) [inline pkg/controller/perconaservermongodb/smart.go:6]: goimport file pls)

File: pkg/psmdb/backup/backup.go [hunk 4]
```diff
@@ -60,13 +74,29 @@ func HasActiveJobs(cl client.Client, cluster, namespace string, current Job) (bo
...
 			return true, nil
 		}
 	}
 
+	pbm, err := NewPBM(cl, cluster)
+	if err != nil {
+		return false, errors.Wrap(err, "getting pbm object")
+	}
+
+	allowLock = append([]LockHeaderPredicate{NotJobLock(current)}, allowLock...)
```
Comment (qJkee) [inline pkg/psmdb/backup/backup.go:90]: could be simplified as
`allowLock = append(allowLock,NotJobLock(current))`

File: pkg/psmdb/backup/pbm.go [hunk 5]
```diff
@@ -146,3 +146,59 @@ func secret(cl client.Client, namespace, secretName string) (*corev1.Secret, err
...
+	allowedByAll := func(l pbm.LockHeader) bool {
+		for _, allow := range predicates {
+			if !allow(l) {
+				return false
+			}
+		}
+		return true
+	}
+
+	for i := range locks {
```
Comment (qJkee) [inline pkg/psmdb/backup/pbm.go:195]:
```
for _,lock := range locks{
..... !allowedByAll(lock.LockHeader)
....
}

File: pkg/controller/perconaservermongodb/smart.go [hunk 6]
```diff
@@ -58,12 +59,21 @@ func (r *ReconcilePerconaServerMongoDB) smartUpdate(cr *api.PerconaServerMongoDB
...
+		return errors.Wrap(err, "failed to check active backups")
 	}
-	if ok {
-		log.Info("can't start 'SmartUpdate': waiting for running backups finished")
+	if isBlocked {
+		log.Info("can't start 'SmartUpdate': waiting for running backups to be finished")
+		return nil
+	}
+
+	isBlocked, err = backup.HasActiveJobs(r.client, cr, backup.Job{}, backup.NotPITRLock)
```
Comment (qJkee) [inline pkg/controller/perconaservermongodb/smart.go:71]: to avoid misleading worth to rename to smth like `hasActiveJobs`

**Review by qJkee — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/smart.go (hunk 3 above)
Comment (ollevche) [reply pkg/controller/perconaservermongodb/smart.go]: updated in 2ab5eebbccd167792aea4b69b5fae6680f19e217

File: pkg/psmdb/backup/backup.go (hunk 4 above)
Comment (ollevche) [reply pkg/psmdb/backup/backup.go]: updated in 2ab5eebbccd167792aea4b69b5fae6680f19e217

File: pkg/psmdb/backup/pbm.go (hunk 5 above)
Comment (ollevche) [reply pkg/psmdb/backup/pbm.go]: updated in 2ab5eebbccd167792aea4b69b5fae6680f19e217

File: pkg/controller/perconaservermongodb/smart.go (hunk 6 above)
Comment (ollevche) [reply pkg/controller/perconaservermongodb/smart.go]: updated in 2ab5eebbccd167792aea4b69b5fae6680f19e217

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | failed |
| version-service | passed |
| upgrade-sharded | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/609/commits/9a865ffc136f63d931dfc8b620d8aa97966eec44
image: `perconalab/percona-server-mongodb-operator:PR-609-9a865ffc`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| upgrade | failed |
| liveness | passed |
| smart-update | passed |
| upgrade-sharded | failed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/609/commits/2ab5eebbccd167792aea4b69b5fae6680f19e217
image: `perconalab/percona-server-mongodb-operator:PR-609-2ab5eebb`

File: pkg/psmdb/backup/pbm.go [hunk 7]
```diff
@@ -146,3 +146,52 @@ func secret(cl client.Client, namespace, secretName string) (*corev1.Secret, err
...
+			if !allow(l) {
+				return false
+			}
+		}
+		return true
+	}
+
+	for _, l := range locks {
+		if !allowedByAll(l.LockHeader) {
+			continue
```
Comment (qJkee) [inline pkg/psmdb/backup/pbm.go:190]: can we make this part is a little bit more readable as a loop above?
Like
```
for _,l := range locks{
  if allowedByAll(l.LockHeader){
    return true,nil
  }
}
return false, nil
```

Comment (ollevche) [reply pkg/psmdb/backup/pbm.go:190]: Yes! Also, I think it should be renamed to `allowedByAllPredicates` cause it's not clear from the function call.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| one-pod | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| rs-shard-migration | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | failed |
| upgrade-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/609/commits/2ab5eebbccd167792aea4b69b5fae6680f19e217
image: `perconalab/percona-server-mongodb-operator:PR-609-2ab5eebb`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/609/commits/3d8fdd310c3f716d09e3fd7a2ebb4b0e4642b230
image: `perconalab/percona-server-mongodb-operator:PR-609-3d8fdd31`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/609/commits/3874b34c75ada358e16832742f48a7c6c9799e6c
image: `perconalab/percona-server-mongodb-operator:PR-609-3874b34c`

**Review by qJkee — APPROVED** [review]

**Pull request merged**